File: pulse-go/internal/pulse/engine/move_list_test.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import (
	"testing"
)

func TestMoveList_add(t *testing.T) {
	t.Run("add should add a move to the move list", func(t *testing.T) {
		ml := &MoveList{}
		m := moveOf(pawnPromotionMove, A7, B8, WhitePawn, BlackQueen, Knight)

		ml.add(m)

		if ml.Size != 1 {
			t.Errorf("move list should have size 1, but got %v", ml.Size)
		}
		if ml.Entries[0].Move != m {
			t.Errorf("first move list entry should have move %v, but got %v", m, ml.Entries[0].Move)
		}
	})
}

func TestMoveList_rateFromMVVLVA(t *testing.T) {
	tests := []struct {
		name string
		move Move
		want value
	}{
		{
			name: "rateByMVVLVA should rate a non-capturing move",
			move: moveOf(normalMove, D4, D5, WhitePawn, NoPiece, NoPieceType),
			want: kingValue / pawnValue,
		},
		{
			name: "rateByMVVLVA should rate a capturing move",
			move: moveOf(normalMove, D1, G4, WhiteQueen, BlackKnight, NoPieceType),
			want: kingValue/queenValue + 10*knightValue,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ml := &MoveList{}
			ml.add(tt.move)
			ml.rateByMVVLVA()
			if ml.Entries[0].value != tt.want {
				t.Errorf("rateByMVVLVA() = %v, want %v", ml.Entries[0].value, tt.want)
			}
		})
	}
}

func TestMoveList_sort(t *testing.T) {
	t.Run("sort should sort all moves", func(t *testing.T) {
		ml := &MoveList{}
		m1 := moveOf(normalMove, D1, G4, WhiteQueen, BlackKnight, NoPieceType)
		m2 := moveOf(normalMove, C1, G5, WhiteBishop, BlackPawn, NoPieceType)
		m3 := moveOf(normalMove, F1, B5, WhiteBishop, NoPiece, NoPieceType)
		m4 := moveOf(normalMove, D4, D5, WhitePawn, NoPiece, NoPieceType)
		ml.add(m4)
		ml.Entries[ml.Size-1].value = 1
		ml.add(m3)
		ml.Entries[ml.Size-1].value = 2
		ml.add(m2)
		ml.Entries[ml.Size-1].value = 3
		ml.add(m1)
		ml.Entries[ml.Size-1].value = 4

		ml.sort()

		if ml.Entries[0].Move != m1 {
			t.Errorf("move at index 0 should be %v, but got %v", m1, ml.Entries[0].Move)
		}
		if ml.Entries[1].Move != m2 {
			t.Errorf("move at index 1 should be %v, but got %v", m2, ml.Entries[1].Move)
		}
		if ml.Entries[2].Move != m3 {
			t.Errorf("move at index 2 should be %v, but got %v", m3, ml.Entries[2].Move)
		}
		if ml.Entries[3].Move != m4 {
			t.Errorf("move at index 3 should be %v, but got %v", m4, ml.Entries[3].Move)
		}
	})
}

File: pulse-kotlin/src/commonTest/kotlin/com/fluxchess/pulse/kotlin/engine/MoveListTest.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

import kotlin.test.Test
import kotlin.test.assertEquals

class MoveListTest {

	@Test
	fun `add should add a move to the move list`() {
		val moveList = MoveList()
		val move = moveOf(PAWN_PROMOTION_MOVE, A7, B8, WHITE_PAWN, BLACK_QUEEN, KNIGHT)

		moveList.add(move)

		assertEquals(1, moveList.size)
		assertEquals(move, moveList.entries[0].move)
	}

	@Test
	fun `rateByMVVLVA should rate a non-capturing move`() {
		val moveList = MoveList()
		moveList.add(moveOf(NORMAL_MOVE, D4, D5, WHITE_PAWN, NO_PIECE, NO_PIECE_TYPE))

		moveList.rateByMVVLVA()

		assertEquals(KING_VALUE / PAWN_VALUE, moveList.entries[0].value)
	}

	@Test
	fun `rateByMVVLVA should rate a capturing move`() {
		val moveList = MoveList()
		moveList.add(moveOf(NORMAL_MOVE, D1, G4, WHITE_QUEEN, BLACK_KNIGHT, NO_PIECE_TYPE))

		moveList.rateByMVVLVA()

		assertEquals(KING_VALUE / QUEEN_VALUE + 10 * KNIGHT_VALUE, moveList.entries[0].value)
	}

	@Test
	fun `sort should sort all moves`() {
		val moveList = MoveList()
		val move1 = moveOf(NORMAL_MOVE, D1, G4, WHITE_QUEEN, BLACK_KNIGHT, NO_PIECE_TYPE)
		val move2 = moveOf(NORMAL_MOVE, C1, G5, WHITE_BISHOP, BLACK_PAWN, NO_PIECE_TYPE)
		val move3 = moveOf(NORMAL_MOVE, F1, B5, WHITE_BISHOP, NO_PIECE, NO_PIECE_TYPE)
		val move4 = moveOf(NORMAL_MOVE, D4, D5, WHITE_PAWN, NO_PIECE, NO_PIECE_TYPE)
		moveList.add(move4)
		moveList.entries[moveList.size - 1].value = 1
		moveList.add(move3)
		moveList.entries[moveList.size - 1].value = 2
		moveList.add(move2)
		moveList.entries[moveList.size - 1].value = 3
		moveList.add(move1)
		moveList.entries[moveList.size - 1].value = 4

		moveList.sort()

		assertEquals(move1, moveList.entries[0].move)
		assertEquals(move2, moveList.entries[1].move)
		assertEquals(move3, moveList.entries[2].move)
		assertEquals(move4, moveList.entries[3].move)
	}
}
